addons/website_sale/tests/test_dynamic_snippet_category.py

```diff
@@ -0,0 +1,91 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from lxml import html
+
+from odoo import Command
+from odoo.addons.website_sale.controllers.main import WebsiteSale
+from odoo.addons.website_sale.tests.common import MockRequest, WebsiteSaleCommon
+
+
+class TestDynamicSnippetCategory(WebsiteSaleCommon):
+    def setUp(self):
+        super().setUp()
+
+        # Create a test product public category
+        self.category1 = self.env['product.public.category'].create({
+            'name': 'Published Category',
+        })
+        self.category2 = self.env['product.public.category'].create({
+            'name': 'Published Category 2',
+        })
+        self.child_category = self.env['product.public.category'].create({
+            'name': 'child category',
+            'parent_id': self.category1.id,
+        })
+        self.category3 = self.env['product.public.category'].create({
+            'name': 'Unpublished Category',
+        })
+        self.env['product.template'].create({
+            'name': 'Test Product',
+            'public_categ_ids': [
+                Command.link(self.category1.id),
+                Command.link(self.category2.id),
+                Command.link(self.child_category.id),
+            ],
+            'website_published': True,
+        })
+        self.website_sale = WebsiteSale()
+        self.website = self.website.with_user(self.env.ref('base.user_admin'))
+
+    def test_get_shop_categories(self):
+        """Test that published categories are returned by the _get_shop_categories"""
+        with MockRequest(self.website.env, website=self.website):
+            categories = self.website_sale.get_shop_categories()
+            root = [html.fromstring(category) for category in categories]
+            self.assertEqual(
+                self.category1.id,
+                int(root[0].attrib.get('data-category-id')),
+                "only published categories should be returned",
+            )
+            self.assertEqual(
+                self.category2.id,
+                int(root[1].attrib.get('data-category-id')),
+                "only published categories should be returned",
+            )
+            categories = self.website_sale.get_shop_categories(self.category1.id)
+            root = [html.fromstring(category) for category in categories]
+            self.assertEqual(
+                self.child_category.id,
+                int(root[0].attrib.get('data-category-id')),
+                "only children categories of category1 should be returned",
+            )
+
+    def test_get_snippet_categories(self):
+        """Test that snippet categories have cover images"""
+        categories = self.env['product.public.category'].get_snippet_categories(self.website.id)
+        self.assertEqual(
+            self.category1.id,
+            categories[0]['id'],
+            "only published categories should be returned",
```

Comment:
```suggestion
            "Only published categories should be returned",
```
Adapt this everywhere